slatedb/src/compactor.rs

```diff
@@ -1915,37 +1881,91 @@ mod tests {
-            Ulid::from_parts(older_ts_ms, 0),
-            CompactionSpec::new(vec![], 10),
+        let compactor_options = Arc::new(compactor_options());
+        let rand = Arc::new(DbRand::default());
+
+        // Seed a stored compaction in the first epoch.
+        let mut handler = CompactorEventHandler::new(
+            manifest_store.clone(),
+            compactions_store.clone(),
+            compactor_options.clone(),
+            Arc::new(MockScheduler::new()),
+            Arc::new(MockExecutor::new()),
+            rand.clone(),
+            Arc::new(CompactionStats::new(Arc::new(StatRegistry::new()))),
+            clock.clone(),
+        )
+        .await
+        .unwrap();
+
+        let first_epoch = handler.state.manifest().value.compactor_epoch;
+        let compaction_id = Ulid::new();
+        let compaction = Compaction::new(
+            compaction_id,
+            CompactionSpec::new(vec![SourceId::Sst(Ulid::new())], 0),
         );
-        let compaction_new = Compaction::new(
-            Ulid::from_parts(newer_ts_ms, 0),
-            CompactionSpec::new(vec![], 11),
+        handler.state.add_compaction(compaction).unwrap();
+        handler.write_compactions_safely().await.unwrap();
+
+        let (_, persisted) = compactions_store
+            .try_read_latest_compactions()
+            .await
+            .unwrap()
+            .expect("compactions should exist after first write");
+        assert!(
+            persisted.iter().next().is_some(),
+            "expected stored compactions to include the seeded compaction"
         );
 
-        fixture
-            .handler
-            .state
-            .add_compaction(compaction_old)
-            .expect("failed to add old compaction");
-        fixture
-            .handler
-            .state
-            .add_compaction(compaction_new)
-            .expect("failed to add new compaction");
+        drop(handler);
 
-        // Update the metric and verify it matches the oldest compaction's ULID time
-        fixture.handler.update_compaction_low_watermark();
+        // Restart compactor (new epoch) and ensure state starts empty.
+        let mut handler = CompactorEventHandler::new(
+            manifest_store.clone(),
+            compactions_store.clone(),
+            compactor_options.clone(),
+            Arc::new(MockScheduler::new()),
+            Arc::new(MockExecutor::new()),
+            rand,
+            Arc::new(CompactionStats::new(Arc::new(StatRegistry::new()))),
+            clock.clone(),
+        )
+        .await
+        .unwrap();
+
+        assert!(handler.state.compactions().next().is_none());
+        assert!(
+            handler.state.manifest().value.compactor_epoch > first_epoch,
+            "compactor epoch should advance on restart"
+        );
+
+        // Persisting after restart should keep only the retained finished compaction for GC.
+        handler.write_compactions_safely().await.unwrap();
```

Comment:
I'm scanning this trying to figure out where we registered the finished compaction?